### ISSUE TITLE
Router

### ISSUE THREAD

**Issue by digitalsadhu** [ash-framework/roadmap#1]
### TODO

Initial phase
- [x] move base router class to ash-framework/classes 
- [x] move base route class to ash-framework/classes 
- [x] do nothing if no index.js route is available
- [x] throw error if route is not available for defined route
- [x] Verify router behaviour is same (similar as appropriate) as ember
- [x] Improve API for Route subclasses
- [ ] Populate `routeName` on route with `get routeName`
- [ ] Verify uniqueness of route params eg. `/users/:user_id/customers/:user_id` would not be valid.
- [x] Add serialize hook.
- [x] Add deserialize hook.
- [ ] flesh out integration tests (all existing tests are integration)
- [ ] Refactor into smaller units
- [ ] Add unit tests (currently tests are very integrated)
- [ ] Flesh out README.md
- [ ] Publish
- [x] Refactor to support file naming conventions for verbs as per @pnw comment below

### API For Route hooks (in order of execution)

``` js
deserialize () {

}
beforeModel () {

}
model () {

}
afterModel (model) {

}
serialize (model) {

}
```

### Properties available on route instance

```js
this.params
this.body
this.query
this.httpContext
```

**Comment by pnw** [comment]
Example router.js

``` js
this.route('users', function () {
  this.route('users', {path: '/'})  // explicit index with custom name - /users
  this.route('user', {path: '/:user_id'}, function () {
    this.route('user', {path: '/'})  // explicit index with custom name - /users/1
    this.route('widgets', {path: '/widgets'}, function () {
      this.route('widgets', {path: '/'})  // explicit index with custom name - users/1/widgets
      this.route('widget', {path: '/:widget_id'})  // users/1/widgets/2
    })
  })
})

this.route('widgets', function () {
  // implicit index - /widgets
  this.route('widget', {path: '/widget_id'})  // /widgets/1
})
```

Corresponding file structure

```
routes/
  users/
    users.get.js
    users.post.js
    user/
      user.get.js
      user.put.js
      user.delete.js
      widgets/
        widgets.get.js
        widgets.post.js
        widget.get.js
        widget.put.js
        widget.delete.js
  widgets/
    index.get.js
    index.post.js
    widget.get.js
    widget.put.js
    widget.delete.js
```

**Comment by pnw** [comment]
Conflicts:

## 1.  Manually overridden index routes

Copied from my comment here: https://github.com/ash-framework/router/pull/1

``` js
// router.js
this.route('users', function () {
  this.route('users', {path: '/'})
})
```

and files like this?

```
routes/
  users/
    index.get.js
    users.get.js
```

In that example, I manually name the "index" path to users, but in the routes directory I have both a users.get.js and an index.get.js.

My feeling is that if I manually override the index route and set this.route('users', {path: '/'}), then the index.get.js file should be ignored.

## 2. Adding the first child route, or how/where is the parent route defined

### tl;dr - ash generate route must always create the route files under a route-directory

### This would not work

`ash generate route users` creates the following

``` js
// router.js
this.route('users')
```

```
routes/
  users.get.js
```

A subsequent call to `ash generate route users/user` creates the following

``` js
// router.js
this.route('users', function () {
  this.route('user', {path: '/:user_id'})  // the path bit isn't auto-generated by ash-cli, but bear with me
})
```

```
routes/
  users.get.js
  users/
    // normally there would be an implicit index.get.js here
    user.get.js
```

_Why this doesn't work:_ we now have three different possible route-files for the `/users` route:
1. `routes/users.get.js`
2. `routes/users/index.get.js` // implicit, automatic behavior
3. `routes/users/ers.get.js` // if the user manually overrides the index path

For this reason, I suggest that generating a new route always creates 1. a directory with the name/path of the route 2. the "index" route within the new directory that was created and not in the parent directory.

Without this rule, you automatically create a conflicting route structure every time you add a child route.

_This would work: Example behavior_

1 - `ash generate route users` 

``` js
// router.js
this.route('users')
```

```
routes/
  users/
    index.get.js
```

2 - `ash generate route users/user --path="/:user_id"`

``` js
// router.js
this.route('users', function () {
  this.route('user', {path: '/:user_id'})
})
```

```
router/
  users/
    index.get.js
    user/
      index.get.js
```

This conflict is born from the fact that the Ash routes behave slightly different from Ember routes.

In Ember, visiting `/users/1/` loads up a route hierarchy in the following form:
1. Application
2. Users
3. User

Whereas in Ash, only a single Route is loaded when you request `/users/1`
1. User

**Comment by digitalsadhu** [comment]
Nice. Will check and adjust.
